adcc/ElectronicTransition.py

```diff
@@ -0,0 +1,269 @@
+        else:
+            raise ValueError("Unknown xaxis specifier: {}".format(xaxis))
+
+        if yaxis in ["osc", "osc_strength", "oscillator_strength", "f"]:
+            absorption = self.oscillator_strength
+            ylabel = "Oscillator strengths (au)"
+        elif yaxis in ["dipole", "dipole_norm", "μ"]:
+            absorption = np.linalg.norm(self.transition_dipole_moment, axis=1)
+            ylabel = "Modulus of transition dipole (au)"
+        elif yaxis in ["cross_section", "σ"]:
+            absorption = self.cross_section
+            ylabel = "Cross section (au)"
+        elif yaxis in ["rot", "rotational_strength", "rotatory_strength"]:
+            absorption = self.rotatory_strength
+            ylabel = "Rotatory strength (au)"
+        else:
+            raise ValueError("Unknown yaxis specifier: {}".format(yaxis))
+
+        sp = ExcitationSpectrum(energies, absorption)
+        sp.xlabel = xlabel
+        sp.ylabel = ylabel
+        if not broadening:
+            plots = sp.plot(style="discrete", **kwargs)
+        else:
+            kwdisc = kwargs.copy()
+            kwdisc.pop("label", "")
+            plots = sp.plot(style="discrete", **kwdisc)
+
+            kwargs.pop("color", "")
+            sp_broad = sp.broaden_lines(width, shape=broadening)
+            plots.extend(sp_broad.plot(color=plots[0].get_color(),
+                                       style="continuous", **kwargs))
+
+        if xaxis in ["nm"]:
+            # Invert x axis
+            plt.xlim(plt.xlim()[::-1])
+        return plots
+
+
+class State2StateTransition(ElectronicTransition):
+    """
+    Documentation
+    """
+    def __init__(self, parent_state, initial=0, final=None):
+        self.parent_state = parent_state
+        self.reference_state = self.parent_state.reference_state
+        self.ground_state = self.parent_state.ground_state
+        self.matrix = self.parent_state.matrix
+        self.property_method = self.parent_state.property_method
+        self.operators = self.parent_state.operators
+
+        self.initial = initial
+        self.final = final
+
+        self.initial_excitation_vector = \
+            self.parent_state.excitation_vector[self.initial]
+        if self.final is None:
+            other_excitation_energy = np.delete(
+                self.parent_state.excitation_energy.copy(), self.initial
+            )
+        else:
+            other_excitation_energy = \
+                np.array([self.parent_state.excitation_energy[self.final]])
+        self.excitation_energy = other_excitation_energy -\
+            self.parent_state.excitation_energy[self.initial]
+    # TODO: describe?!
+
+    @cached_property
+    @mark_excitation_property(transform_to_ao=True)
+    @timed_member_call(timer="_property_timer")
+    def transition_dm(self):
+        """
+        List of transition density matrices from
+        initial state to final state/s
+        """
+        # TODO: only states above self.initial
```

Comment:
Yes. Otherwise I think the interface becomes very strange.

Reply:
Ok this interface is very confusing (at least for me). This initial final business we need to talk about at some point.

Reply:
I think the main problem is:

Do we allow for only one or multiple final states in State2States? Or both?

From my point of view, we could make it lazy, such that you can either get all the transitions from one initial state to all the others...

Reply:
That's a good question. In my head it was one initial to all final (kind of as a pendant to ExcitedStates). But you probably have more insight into what is needed. I think it gets pretty tricky in the interfaces to support both at the same time. What one could do is have one class that does 1 inital -> all final and one that does the reverse (all initial -> 1 final). Or one that does truely everything (all -> all) and then some clever iterators or arrangement of the properties in multi-dimensional np arrays to make it easy to use the results. But I don't know how expensive that becomes. Do you have an idea about that?

Reply:
I think 1 initial -> all final is the best solution. Then it's most consistent with `ExcitedStates`. all -> all is probably too expensive by default...

So I'm going to adapt to 1 initial -> all final and then we can decide whether we like it or not 😄

Reply:
1 initial -> all final now implemented.